src/System.Net.Requests/src/System/Net/_FtpControlStream.cs

```diff
@@ -0,0 +1,1299 @@
+                {
+                    string baseDir = (requestDirectory == string.Empty) 
+                        ? string.Empty : requestDirectory + "/";
+                    commandList.Add(new PipelineEntry(FormatFtpCommand("RNFR", baseDir + requestFilename), flags));
+
+                    string renameTo;
+                    if (!string.IsNullOrEmpty(request.RenameTo)
+                        && request.RenameTo.StartsWith("/", StringComparison.OrdinalIgnoreCase))
+                    {
+                        renameTo = request.RenameTo; // Absolute path
+                    }
+                    else
+                    {
+                        renameTo = baseDir + request.RenameTo; // Relative path
+                    }
+                    commandList.Add(new PipelineEntry(FormatFtpCommand("RNTO", renameTo), flags));
+                }
+                else if (request.MethodInfo.HasFlag(FtpMethodFlags.DoesNotTakeParameter))
+                {
+                    commandList.Add(new PipelineEntry(FormatFtpCommand(request.Method, string.Empty), flags));
+                }
+                else if (request.MethodInfo.HasFlag(FtpMethodFlags.MustChangeWorkingDirectoryToPath))
+                {
+                    commandList.Add(new PipelineEntry(FormatFtpCommand(request.Method, requestFilename), flags));
+                }
+                else
+                {
+                    commandList.Add(new PipelineEntry(FormatFtpCommand(request.Method, requestPath), flags));
+                }
+            
+                if (!request.KeepAlive)
+                {
+                    commandList.Add(new PipelineEntry(FormatFtpCommand("QUIT", null)));
+                }
+#if FTP_FEATURE_CACHE
+            }
+#endif
+
+            return (PipelineEntry []) commandList.ToArray(typeof(PipelineEntry));
+        }
+
+        private PipelineInstruction QueueOrCreateDataConection(PipelineEntry entry, ResponseDescription response, bool timeout, ref Stream stream, out bool isSocketReady)
+        {
+            isSocketReady = false;
+            if (m_DataHandshakeStarted)
+            {
+                isSocketReady = true;
+                return PipelineInstruction.Pause; //if we already started then this is re-entering into the callback where we proceed with the stream
+            }
+
+            m_DataHandshakeStarted = true;
+
+            // handle passive responses by parsing the port and later doing a Connect(...)
+            bool isPassive = false;
+            int port = -1;
+            if (entry.Command == "PASV\r\n" || entry.Command == "EPSV\r\n")
+            {
+                if (!response.PositiveCompletion)
+                {
+                    m_AbortReason = SR.Format(SR.net_ftp_server_failed_passive, response.Status);
+                    return PipelineInstruction.Abort;
+                }
+                if (entry.Command == "PASV\r\n")
```

Comment:
I assume the protocol dictates \r\n and this isn't meant to be a platform-specific newline, right?